### DISCUSSION TITLE
Knex/PostgreSQL error: "sorry, too many clients already"

### DISCUSSION THREAD

**Discussion by vrilcode** [lowdefy/lowdefy#1472] [Questions and Help]
When I use Knex with PostgreSQL, connections are not reused (pooling) nor released. Connections are used for one query only and then persist as idle. After some further Lowdefy requests, used DB connections are growing till all are taken. How long it takes  depends on the PostgreSQL connection limit. My PG connection limit is 100, Knex pool is set to min 0 and max 10. According to Knex docs idle connections should timeout in 30 secs, but that never happens. When I stop `npx lowdefy@latest dev` all connections are released. So after a couple of requests, I have to restart Lowdefy backend.

Error message:
```
RequestError: sorry, too many clients already
    at Ue.callResolver (/Users/vrilcode/my_project/.lowdefy/.cache/scripts/graphql/3_23_3/package/dist/253.index.js:85:107095)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Ue.callRequest (/Users/vrilcode/my_project/.lowdefy/.cache/scripts/graphql/3_23_3/package/dist/253.index.js:85:104850)
```
Is it possible, that Lowdefy initiates Knex library multiple times, eg. on each request?
At the moment I haven't found any other explanation.

[Knex docs](https://knexjs.org/guide/#configuration-options): "Initializing the library should normally only ever happen once in your application, as it creates a connection pool for the current database, you should use the instance returned from the initialize call throughout your library."

**Comment by vrilcode** [comment]
I haven't checked it, but this could be the cause:

* https://github.com/lowdefy/lowdefy/blob/aede980a615d339db496a423820612ef85bcbebf/packages/graphql/src/connections/Knex/KnexBuilder/KnexBuilder.js#L22
* https://github.com/lowdefy/lowdefy/blob/aede980a615d339db496a423820612ef85bcbebf/packages/graphql/src/connections/Knex/KnexRaw/KnexRaw.js#L21

I would check it and fix it, but I don't know what the preferred way in Lowdefy for holding global data is.

**Comment by vrilcode** [comment]
At the moment it works as it should (connection idle timeout 30 secs).  Postgresql and Knex configuration was not modified. I just changed some KnexRaw requests to KnexBuilder requests. Very strange. I will give an update, if it occurs again.

**Reply by vicam001** [reply]
hi @vrilcode , I suggest you have a look at https://github.com/lowdefy/lowdefy/discussions/1370, I faced similar challenges and my solution was to add a proxy. In different applications I have used AWS RDS Proxy, Supabase or Hasura (in this case because the requirements pushed me towards GraphQL queries) and I no longer hit the limit of database connections.

**Reply by vrilcode** [reply]
Thanks for you hint. I've seen your thread and such a proxy might be a good workaround as long as the problem is unsolved. Nevertheless it should work without proxy. When v4 is released, I will examine the connection plugin.
